### ISSUE TITLE
readers.las: access via HTTPS/S3 downloads the file twice

### ISSUE THREAD

**Issue by rcoup** [PDAL/PDAL#3847]
**Describe the bug**

HTTPS/S3 access to a non-COPC LAS/LAZ files via `pdal info` or an equivalent pipeline does two complete file downloads.

```console
$ VERBOSE=1 build/bin/pdal -v 8 info --schema --metadata --driver=readers.las "https://media.githubusercontent.com/media/PDAL/data/master/autzen/autzen.laz"
...
> GET /media/PDAL/data/master/autzen/autzen.laz HTTP/2
> HEAD /media/PDAL/data/master/autzen/autzen.laz HTTP/2
> GET /media/PDAL/data/master/autzen/autzen.laz HTTP/2
> HEAD /media/PDAL/data/master/autzen/autzen.laz HTTP/2
```

I understand readers.las doesn't have the ability to do partial range reads over HTTP, but that's a separate issue.

See more detailed debugging at https://github.com/PDAL/PDAL/issues/3846

**Expected behavior**

Does a single download.

**System/installation information:**

```
$ uname -a
Darwin inchgarvie.local 21.6.0 Darwin Kernel Version 21.6.0: Sat Jun 18 17:07:22 PDT 2022; root:xnu-8020.140.41~1/RELEASE_ARM64_T6000 arm64 arm Darwin

$ pdal --version
pdal 2.5.0 (git-version: 6cb54d)
```

**Comment by hobu** [comment]
Not excited to put any kind of caching machinery in Arbiter...

**Comment by abellgithub** [comment]
I think this has come up before and people have PR'd various caching schemes or other fixes, but they all have downsides and add complication. I suppose you could add some sort of very basic cache control where files attached to a stage could go away when the stage was destroyed, but even this takes some work and it's hard know if the benefit is worth the complexity when one can just download the file and work locally.

I know that there was also some talk about using fancier remote access code like GDAL has or whatever, but again, one has to ask if the benefit is worth the complexity.

**Comment by rcoup** [comment]
> Not excited to put any kind of caching machinery in Arbiter...

Sure. Maybe an API to indicate the file is 100% going to be reopened, and save removal for the final access/pipeline end/process exit? Something vaguely like: `FileUtils::openFile(path, hints=LATER_READ)` providing a hint to the file access layer (Arbiter) that this path will very likely be reopened again in the same pipeline/process for further reading. For local filesystem access it can obviously be ignored.

---

_Personally_, I think a better approach is to call `setrlimit()` to up the FD limit to something appropriate, then open files once for their lifetime of use. And if that's not enough, fail cleanly with help for the user to get it fixed. Hard limits are circa 1M descriptors on most Linux distros by default, unlimited on macOS, 16M on Windows. etc.

**Comment by rcoup** [comment]
My underlying driver is to read the header info, PDRF, point counts, CRS, extent _without_ downloading the whole file. This was just a bug I encountered along the way :-)

**Comment by hobu** [comment]
Implemented in #3900
